### PR TITLE
added analyzer for conflicting gateways based on issue#27481

### PR DESCRIPTION
**Pull request by zhlsunshine**
This PR is for conflicting gateways based on #27481. 
1. Add analyzer for conflicting gateway analysis
2. Add new analyzer unit test

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[x] Test and Release
[x] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[ ] Does not have any changes that may affect Istio users.

### CODE REVIEW

**Comment by zhlsunshine** [comment]
Here is functional verification below:

**Step1:** created 4 conflicting gateways `alpha`, `beta`, `theta` and `gama`, showing as below:
```
$ kubectl get gateway
NAME                  AGE
alpha                   9s
beta                     9s
gama                   9s
theta                    9s
```

**Step 2:**  execute analysis by `istioctl analyze`, the result shows below:
```
$ istioctl analyze -A
Error [IST0145] (Gateway alpha.default) Conflicting gateways are detected for gateways default/alpha to default/gama,default/theta,default/beta on selector, port number and hosts.
Error [IST0145] (Gateway beta.default) Conflicting gateways are detected for gateways default/beta to default/gama,default/theta,default/alpha on selector, port number and hosts.
Error [IST0145] (Gateway gama.default) Conflicting gateways are detected for gateways default/gama to default/alpha,default/theta,default/beta on selector, port number and hosts.
Error [IST0145] (Gateway theta.default) Conflicting gateways are detected for gateways default/theta to default/beta,default/alpha,default/gama on selector, port number and hosts.
Warning [IST0144] (Deployment istio-egressgateway.istio-system) User ID (UID) 1337 is reserved for the sidecar proxy.
Warning [IST0144] (Deployment istio-ingressgateway.istio-system) User ID (UID) 1337 is reserved for the sidecar proxy.
Warning [IST0144] (Deployment istiod.istio-system) User ID (UID) 1337 is reserved for the sidecar proxy.
Warning [IST0144] (Pod istio-egressgateway-88579fc49-9gmnx.istio-system) User ID (UID) 1337 is reserved for the sidecar proxy.
Warning [IST0144] (Pod istio-ingressgateway-664bdff874-gsjw6.istio-system) User ID (UID) 1337 is reserved for the sidecar proxy.
Warning [IST0144] (Pod istiod-67fc95944d-66psr.istio-system) User ID (UID) 1337 is reserved for the sidecar proxy.
Info [IST0102] (Namespace calico-system) The namespace is not enabled for Istio injection. Run 'kubectl label namespace calico-system istio-injection=enabled' to enable it, or 'kubectl label namespace calico-system istio-injection=disabled' to explicitly mark it as not needing injection.
Info [IST0102] (Namespace istio-system) The namespace is not enabled for Istio injection. Run 'kubectl label namespace istio-system istio-injection=enabled' to enable it, or 'kubectl label namespace istio-system istio-injection=disabled' to explicitly mark it as not needing injection.
Info [IST0102] (Namespace tigera-operator) The namespace is not enabled for Istio injection. Run 'kubectl label namespace tigera-operator istio-injection=enabled' to enable it, or 'kubectl label namespace tigera-operator istio-injection=disabled' to explicitly mark it as not needing injection.
Info [IST0118] (Service calico-typha.calico-system) Port name calico-typha (port: 5473, targetPort: calico-typha) doesn't follow the naming convention of Istio port.
Error: Analyzers found issues when analyzing all namespaces.
See https://istio.io/v1.10/docs/reference/config/analysis for more information about causes and resolutions.
```

**Step3:** confirm the unit test based on [issue#27481](https://github.com/istio/istio/issues/27481) and result shows below:
$make test
```
......
ok  	istio.io/istio/cni/test	(cached)
ok  	istio.io/istio/galley/pkg/config/analysis	(cached)
ok  	istio.io/istio/galley/pkg/config/analysis/analyzers	6.547s
?   	istio.io/istio/galley/pkg/config/analysis/analyzers/annotations	[no test files]
ok  	istio.io/istio/galley/pkg/config/analysis/analyzers/authz	(cached)
?   	istio.io/istio/galley/pkg/config/analysis/analyzers/deployment	[no test files]
?   	istio.io/istio/galley/pkg/config/analysis/analyzers/deprecation	[no test files]
?   	istio.io/istio/galley/pkg/config/analysis/analyzers/destinationrule	[no test files]
?   	istio.io/istio/galley/pkg/config/analysis/analyzers/gateway	[no test files]
?   	istio.io/istio/galley/pkg/config/analysis/analyzers/injection	[no test files]
?   	istio.io/istio/galley/pkg/config/analysis/analyzers/multicluster	[no test files]
ok  	istio.io/istio/galley/pkg/config/analysis/analyzers/schema	(cached)
?   	istio.io/istio/galley/pkg/config/analysis/analyzers/service	[no test files]
?   	istio.io/istio/galley/pkg/config/analysis/analyzers/serviceentry	[no test files]
?   	istio.io/istio/galley/pkg/config/analysis/analyzers/sidecar	[no test files]
ok  	istio.io/istio/galley/pkg/config/analysis/analyzers/util	(cached)
?   	istio.io/istio/galley/pkg/config/analysis/analyzers/virtualservice	[no test files]
?   	istio.io/istio/galley/pkg/config/analysis/analyzers/webhook	[no test files]
ok  	istio.io/istio/galley/pkg/config/analysis/diag	(cached)
ok  	istio.io/istio/galley/pkg/config/analysis/local	(cached)
?   	istio.io/istio/galley/pkg/config/analysis/msg	[no test files]
?   	istio.io/istio/galley/pkg/config/analysis/testing/fixtures	[no test files]
ok  	istio.io/istio/galley/pkg/config/collection	(cached)
......
```

File: galley/pkg/config/analysis/analyzers/gateway/conflictinggateway.go
Comment (shamsher31) [inline galley/pkg/config/analysis/analyzers/gateway/conflictinggateway.go:22]: nit: remove comment
```suggestion

```

File: galley/pkg/config/analysis/msg/messages.yaml
Comment (shamsher31) [inline galley/pkg/config/analysis/msg/messages.yaml:461]:
```suggestion
    template: "Conflicting gateways are detected for gateway %s on selector, port number and hosts."
```

File: galley/pkg/config/analysis/analyzers/gateway/conflictinggateway.go
Comment (shamsher31) [inline galley/pkg/config/analysis/analyzers/gateway/conflictinggateway.go:112]: Let's keep this func private if it's not going to be used outside this analyzer.

File: galley/pkg/config/analysis/msg/messages.yaml
Comment (shamsher31) [inline galley/pkg/config/analysis/msg/messages.yaml:461]: It will be good to specify selector, port number, and host in the description?

```
Error [IST0145] (Gateway alpha.default) Conflicting gateways are detected for gateway default/alpha to default/gama,default/theta,default/beta
on selector 'ingressgateway' with port number '80' and hosts 'foo.bar'.
```

File: galley/pkg/config/analysis/analyzers/gateway/conflictinggateway.go
Comment (zhlsunshine) [reply galley/pkg/config/analysis/analyzers/gateway/conflictinggateway.go:22]: okay, thanks, I will remove it.

File: galley/pkg/config/analysis/msg/messages.yaml
Comment (zhlsunshine) [reply galley/pkg/config/analysis/msg/messages.yaml:461]: okay, I will change it to make them to be unified

File: galley/pkg/config/analysis/analyzers/gateway/conflictinggateway.go
Comment (zhlsunshine) [reply galley/pkg/config/analysis/analyzers/gateway/conflictinggateway.go:112]: okay, I will make it invisibility outside of this analyzer

File: galley/pkg/config/analysis/msg/messages.yaml
Comment (zhlsunshine) [reply galley/pkg/config/analysis/msg/messages.yaml:461]: okay, that's great to make this error information more useful

**Comment by zhlsunshine** [comment]
Hi @shamsher31, all code changes are done, thank you for code reviewing.  And here is the analysis result:

```
$ istioctl analyze -A
Error [IST0145] (Gateway alpha.default) Conflicting gateways are detected for gateway default/alpha to default/gama,default/theta,default/beta on selector 'istio=ingressgateway', port number '80' and hosts '[foo.bar]'.
Error [IST0145] (Gateway beta.default) Conflicting gateways are detected for gateway default/beta to default/alpha,default/gama,default/theta on selector 'istio=ingressgateway', port number '80' and hosts '[foo.bar]'.
Error [IST0145] (Gateway gama.default) Conflicting gateways are detected for gateway default/gama to default/theta,default/beta,default/alpha on selector 'istio=ingressgateway', port number '80' and hosts '[foo.bar]'.
Error [IST0145] (Gateway theta.default) Conflicting gateways are detected for gateway default/theta to default/gama,default/beta,default/alpha on selector 'istio=ingressgateway', port number '80' and hosts '[foo.bar]'.
Warning [IST0144] (Deployment istio-egressgateway.istio-system) User ID (UID) 1337 is reserved for the sidecar proxy.
Warning [IST0144] (Deployment istio-ingressgateway.istio-system) User ID (UID) 1337 is reserved for the sidecar proxy.
Warning [IST0144] (Deployment istiod.istio-system) User ID (UID) 1337 is reserved for the sidecar proxy.
Warning [IST0144] (Pod istio-egressgateway-88579fc49-5wnhz.istio-system) User ID (UID) 1337 is reserved for the sidecar proxy.
Warning [IST0144] (Pod istio-ingressgateway-664bdff874-gx6vs.istio-system) User ID (UID) 1337 is reserved for the sidecar proxy.
Warning [IST0144] (Pod istiod-67fc95944d-zsn6c.istio-system) User ID (UID) 1337 is reserved for the sidecar proxy.
Info [IST0102] (Namespace calico-system) The namespace is not enabled for Istio injection. Run 'kubectl label namespace calico-system istio-injection=enabled' to enable it, or 'kubectl label namespace calico-system istio-injection=disabled' to explicitly mark it as not needing injection.
Info [IST0102] (Namespace istio-system) The namespace is not enabled for Istio injection. Run 'kubectl label namespace istio-system istio-injection=enabled' to enable it, or 'kubectl label namespace istio-system istio-injection=disabled' to explicitly mark it as not needing injection.
Info [IST0102] (Namespace tigera-operator) The namespace is not enabled for Istio injection. Run 'kubectl label namespace tigera-operator istio-injection=enabled' to enable it, or 'kubectl label namespace tigera-operator istio-injection=disabled' to explicitly mark it as not needing injection.
Info [IST0118] (Service calico-typha.calico-system) Port name calico-typha (port: 5473, targetPort: calico-typha) doesn't follow the naming convention of Istio port.
Error: Analyzers found issues when analyzing all namespaces.
See https://istio.io/v1.10/docs/reference/config/analysis for more information about causes and resolutions.
```

**Comment by zhlsunshine** [comment]
Hi @shamsher31 done for all code change. Please help to review again. Thanks! ^_^

**Comment by zhlsunshine** [comment]
Hi @istio/wg-networking-maintainers , I have to change the integration unit test for completing this new analyzer, because there are gateways errors have been detected by new analyzer in such unit test. Could you please help to approve it? Thanks!

**Comment by irisdingbj** [comment]
@therealmitchconnors @esnible PTAL for the newly added analyzer, Thanks!

**Review by shamsher31 — COMMENTED** [review]
Overall LGTM. I will let @esnible take a another look.

**Review by esnible — COMMENTED** [review]
Looks promising.

The message `Error [IST0145] (Gateway alpha.default) Conflicting gateways are detected for gateway default/alpha to default/gama,default/theta,default/beta on selector 'istio=ingressgateway', port number '80' and hosts '[foo.bar]'.` mentions the gateway twice, ones in the parenthesis and once in the text.

It is also strange to see the port number quoted and the hosts bracketed.  Perhaps the message should look like

```
Error [IST0145] (Gateway alpha.default) Conflict with gateways default/gama, default/theta, default/beta (workload selector istio=ingressgateway, port 80, hosts foo.bar.
```

**Comment by zhlsunshine** [comment]
> Looks promising.
> 
> The message `Error [IST0145] (Gateway alpha.default) Conflicting gateways are detected for gateway default/alpha to default/gama,default/theta,default/beta on selector 'istio=ingressgateway', port number '80' and hosts '[foo.bar]'.` mentions the gateway twice, ones in the parenthesis and once in the text.
> 
> It is also strange to see the port number quoted and the hosts bracketed. Perhaps the message should look like
> 
> ```
> Error [IST0145] (Gateway alpha.default) Conflict with gateways default/gama, default/theta, 
> default/beta (workload selector istio=ingressgateway, port 80, hosts foo.bar).
> ```

Hi @esnible ,  thank you for your time and reply on this PR, I have changed the code accordingly, here is the analysis result:

```
$ istioctl analyze -A
Error [IST0145] (Gateway alpha.default) Conflict with gateways default/theta,default/beta,default/gama (workload selector istio=ingressgateway, port 80, hosts foo.bar).
Error [IST0145] (Gateway beta.default) Conflict with gateways default/theta,default/alpha,default/gama (workload selector istio=ingressgateway, port 80, hosts foo.bar).
Error [IST0145] (Gateway gama.default) Conflict with gateways default/theta,default/beta,default/alpha (workload selector istio=ingressgateway, port 80, hosts foo.bar).
Error [IST0145] (Gateway theta.default) Conflict with gateways default/beta,default/alpha,default/gama (workload selector istio=ingressgateway, port 80, hosts foo.bar).
Warning [IST0144] (Deployment istio-egressgateway.istio-system) User ID (UID) 1337 is reserved for the sidecar proxy.
Warning [IST0144] (Deployment istio-ingressgateway.istio-system) User ID (UID) 1337 is reserved for the sidecar proxy.
Warning [IST0144] (Deployment istiod.istio-system) User ID (UID) 1337 is reserved for the sidecar proxy.
Warning [IST0144] (Pod istio-egressgateway-88579fc49-m7vlq.istio-system) User ID (UID) 1337 is reserved for the sidecar proxy.
Warning [IST0144] (Pod istio-ingressgateway-664bdff874-qhx9h.istio-system) User ID (UID) 1337 is reserved for the sidecar proxy.
Warning [IST0144] (Pod istiod-67fc95944d-s9k5x.istio-system) User ID (UID) 1337 is reserved for the sidecar proxy.
Info [IST0102] (Namespace calico-system) The namespace is not enabled for Istio injection. Run 'kubectl label namespace calico-system istio-injection=enabled' to enable it, or 'kubectl label namespace calico-system istio-injection=disabled' to explicitly mark it as not needing injection.
Info [IST0102] (Namespace istio-system) The namespace is not enabled for Istio injection. Run 'kubectl label namespace istio-system istio-injection=enabled' to enable it, or 'kubectl label namespace istio-system istio-injection=disabled' to explicitly mark it as not needing injection.
Info [IST0102] (Namespace tigera-operator) The namespace is not enabled for Istio injection. Run 'kubectl label namespace tigera-operator istio-injection=enabled' to enable it, or 'kubectl label namespace tigera-operator istio-injection=disabled' to explicitly mark it as not needing injection.
Info [IST0118] (Service calico-typha.calico-system) Port name calico-typha (port: 5473, targetPort: calico-typha) doesn't follow the naming convention of Istio port.
Error: Analyzers found issues when analyzing all namespaces.
See https://istio.io/v1.10/docs/reference/config/analysis for more information about causes and resolutions.
```
Thanks again! ^_^

**Comment by zhlsunshine** [comment]
Hi @esnible & @shamsher31 ,  could you please help to approve this PR when you are available if it's okay? Thanks! ^_^

**Review by hzxuzhonghu — APPROVED** [review]
this is just one special case, i think no harm to add first.

**Review by shamsher31 — APPROVED** [review]
@zhlsunshine works great. Thanks for your PR. 
```
$ ./out/linux_amd64/istioctl analyze galley/pkg/config/analysis/analyzers/testdata/conflicting-gateways.yaml
Error [IST0145] (Gateway alpha.default galley/pkg/config/analysis/analyzers/testdata/conflicting-gateways.yaml:16) Conflict with gateways default/beta (workload selector istio=ingressgateway, port 80, hosts foo.bar).
Error [IST0145] (Gateway beta.default galley/pkg/config/analysis/analyzers/testdata/conflicting-gateways.yaml:1) Conflict with gateways default/alpha (workload selector istio=ingressgateway, port 80, hosts foo.bar).
Info [IST0102] (Namespace default) The namespace is not enabled for Istio injection. Run 'kubectl label namespace default istio-injection=enabled' to enable it, or 'kubectl label namespace default istio-injection=disabled' to explicitly mark it as not needing injection.
Error: Analyzers found issues when analyzing namespace: default.
See https://istio.io/v1.10/docs/reference/config/analysis for more information about causes and resolutions.
```

**Pull request merged**
